lib/go-atscfg/meta.go

```diff
@@ -0,0 +1,195 @@
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import (
+	"encoding/json"
+	"strings"
+
+	"github.com/apache/trafficcontrol/lib/go-log"
+	"github.com/apache/trafficcontrol/lib/go-tc"
+)
+
+type ConfigProfileParams struct {
+	FileNameOnDisk string
+	Location       string
+	URL            string
+	APIURI         string
+}
+
+func MakeMetaConfig(
+	serverHostName tc.CacheName,
+	server *ServerInfo,
+	tmURL string, // global tm.url Parameter
+	tmReverseProxyURL string, // global tm.rev_proxy.url Parameter
+	locationParams map[string]ConfigProfileParams, // map[configFile]params; 'location' and 'URL' Parameters on serverHostName's Profile
+	uriSignedDSes []tc.DeliveryServiceName,
+	scopeParams map[string]string, // map[configFileName]scopeParam
+) string {
+	if tmURL == "" {
+		log.Errorln("ats.GetConfigMetadata: global tm.url parameter missing or empty! Setting empty in meta config!")
+	}
+
+	atsData := tc.ATSConfigMetaData{
+		Info: tc.ATSConfigMetaDataInfo{
+			ProfileID:         int(server.ProfileID),
+			TOReverseProxyURL: tmReverseProxyURL,
+			TOURL:             tmURL,
+			ServerIPv4:        server.IP,
+			ServerPort:        server.Port,
+			ServerName:        server.HostName,
+			CDNID:             server.CDNID,
+			CDNName:           string(server.CDN),
+			ServerID:          server.ID,
+			ProfileName:       server.ProfileName,
+		},
+		ConfigFiles: []tc.ATSConfigMetaDataConfigFile{},
+	}
+
+	if locationParams["remap.config"].Location != "" {
+		configLocation := locationParams["remap.config"].Location
+		for _, ds := range uriSignedDSes {
+			cfgName := "uri_signing_" + string(ds) + ".config"
+			// If there's already a parameter for it, don't clobber it. The user may wish to override the location.
+			if _, ok := locationParams[cfgName]; !ok {
+				p := locationParams[cfgName]
+				p.FileNameOnDisk = cfgName
+				p.Location = configLocation
+				locationParams[cfgName] = p
+			}
+		}
+	}
+
+	for cfgFile, cfgParams := range locationParams {
+		atsCfg := tc.ATSConfigMetaDataConfigFile{
+			FileNameOnDisk: cfgParams.FileNameOnDisk,
+			Location:       cfgParams.Location,
+		}
+
+		scope := getServerScope(cfgFile, server.Type, scopeParams)
+
+		if cfgParams.URL != "" {
+			scope = tc.ATSConfigMetaDataConfigFileScopeCDNs
+			atsCfg.URL = cfgParams.URL
+		} else {
+			scopeID := ""
+			if scope == tc.ATSConfigMetaDataConfigFileScopeCDNs {
+				scopeID = string(server.CDN)
+			} else if scope == tc.ATSConfigMetaDataConfigFileScopeProfiles {
+				scopeID = server.ProfileName
+			} else { // ATSConfigMetaDataConfigFileScopeServers
+				scopeID = server.HostName
+			}
+			atsCfg.APIURI = "/api/1.2/" + string(scope) + "/" + scopeID + "/configfiles/ats/" + cfgFile
```

Comment:
Hardcoded 1.2?

Reply:
The config generator doesn't actually care what this number is. Unfortunately, the only place it's really defined is in the client, and we don't want this library to depend on that.
I changed it to a const.